### DISCUSSION TITLE
[WASM][WinUI][AOT] Build fails

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#7741] [Q&A]
### Current behavior

The WASM AOT build fails on my custom WinUI project using VS2022 + .NET6 + latest UNO prelease packages. Turning on diagnostics mode does not show more on the actual cause than this:

> 3>    [00:06:28.3952472] ninja: build stopped: subcommand failed.
3>    C:\Users\ddroth\.nuget\packages\uno.wasm.bootstrap\3.1.3\build\Uno.Wasm.Bootstrap.targets(175,5): error : System.Exception: Failed to generate AOT layout (More details are available in diagnostics mode or using the MSBuild /bl switch)
3>    C:\Users\ddroth\.nuget\packages\uno.wasm.bootstrap\3.1.3\build\Uno.Wasm.Bootstrap.targets(175,5): error :    at Uno.Wasm.Bootstrap.ShellTask_va01173706452d6a3642c80df957b3b52c781fe91.RunPackager() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 684
3>    C:\Users\ddroth\.nuget\packages\uno.wasm.bootstrap\3.1.3\build\Uno.Wasm.Bootstrap.targets(175,5): error :    at Uno.Wasm.Bootstrap.ShellTask_va01173706452d6a3642c80df957b3b52c781fe91.Execute() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 225
3>  Done executing task "ShellTask_va01173706452d6a3642c80df957b3b52c781fe91" -- FAILED.

Not sure if relevant, but I have the (unfounded?) suspicion that this problem might be related to the significant size of custom project

### Expected behavior

Building WASM AOT should work

**Comment by jeromelaban** [comment]
To troubleshoot AOT builds, please use the [following instructions](https://platform.uno/docs/articles/uno-builds-troubleshooting.html#generating-msbuild-binary-log-files).

**Reply by DierkDroth** [reply]
I followed these instructions and clicked on that failed step. However, the I doubt the outcome (see attached screenshots) will tell you much

![Untitled](https://user-images.githubusercontent.com/26179242/148105885-20a3686a-8dc3-4ac3-908e-c3db7534adc6.png)

![Untitled2](https://user-images.githubusercontent.com/26179242/148106094-6401d7a8-4a2d-4a65-8e79-0feb8b2fb38f.png)

**Reply by jeromelaban** [reply]
You can open the binlogs using the viewer tool mentioned in the article.

**Reply by DierkDroth** [reply]
Sorry, I don't know how to retrieve more info than what I already provided. I needed more detailed instructions in case you're looking for something different.
Here is what I see when right clicking on the failed step + "Open Logs External..."
![Untitled3](https://user-images.githubusercontent.com/26179242/148110023-17116722-25a9-4f44-bd4a-b896e2f0910a.png)

**Reply by jeromelaban** [reply]
Thanks. This may look like https://github.com/unoplatform/uno/issues/6152.

Try adding:
```
<ItemGroup>
	<MonoRuntimeMixedModeExcludedAssembly Include="Uno.UI.FluentTheme.v1" />
	<MonoRuntimeMixedModeExcludedAssembly Include="Uno.UI.FluentTheme.v2" />
	<MonoRuntimeMixedModeExcludedAssembly Include="SkiaSharp" />
</ItemGroup>
```

to your project.

**Reply by DierkDroth** [reply]
Thanks for your suggestion. However, I didn't help: still same error.

Also:
- I don't have any SkiaSharp in my custom project. So I would have been surprised if that would have made any difference
- I'm only doing standard WindowsAppSDK UI stuff, no additional NuGet packages whatsoever. So I would have been surprised if I had to add lines above which are not needed by the UNO standard template anyway

**Reply by jeromelaban** [reply]
The SkiaSharp line is a copy paste from the last resolution step of the mentioned issue. Try using the first two lines (v1 and v2).

**Reply by DierkDroth** [reply]
I just tried that: same problem, AOT fails to build with same error.

Wouldn't the exception stack trace provide any hint on what might be going wrong?

**Reply by jeromelaban** [reply]
There's likely an error raised by emscripten somewhere in the binlog, for the builddist target. You may search for "error" or "failure" in the outputs for the builddist target.

**Reply by DierkDroth** [reply]
I followed your suggestions. However, the only error I found (binlog attached) was the exact one which I reported above.

What else could be done to isolate the cause of trouble? Isn't the failing feature part of Uno.Wasm.Bootstrap and as such inspecting the exception callstack could provide us a hint?!?

[msbuild.zip](https://github.com/unoplatform/uno/files/7833601/msbuild.zip)

**Comment by jeromelaban** [comment]
This is the error:

```
    [00:03:15.3985762] [730/764] [AOT] .\linker-out\Uno.UI.FluentTheme.v2.dll -> ./Uno.UI.FluentTheme.v2.dll.bc.tmp
    [00:03:15.4003194] FAILED: Uno.UI.FluentTheme.v2.dll.bc.tmp 
```

Make sure that you have set:

```
<ItemGroup>
	<MonoRuntimeMixedModeExcludedAssembly Include="Uno.UI.FluentTheme.v1" />
	<MonoRuntimeMixedModeExcludedAssembly Include="Uno.UI.FluentTheme.v2" />
</ItemGroup>
```

in your project.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban 

I tried that before and just tried it again: still same error (see attached). I also attached my project file.

[msbuild.zip](https://github.com/unoplatform/uno/files/7839448/msbuild.zip)
[Unusual.Wasm.zip](https://github.com/unoplatform/uno/files/7839462/Unusual.Wasm.zip)

**Reply by jeromelaban** [reply]
Thanks, `FullAOT` is not supported in net6 at this time. Use `InterpreterAndAOT` instead.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban. That got me going...
